common/src/main/java/io/netty/util/Recycler.java

```diff
@@ -256,12 +254,17 @@ public void recycle(Object object) {
                 this.availableSharedCapacity = availableSharedCapacity;
             }
 
+            /// TODO: In the future when we move to Java9+ we should use java.lang.ref.Cleaner.
             @Override
-            public void run() {
-                Link head = link;
-                while (head != null) {
-                    reclaimSpace(LINK_CAPACITY);
-                    head = head.next;
+            protected void finalize() throws Throwable {
+                try {
+                    super.finalize();
+                } finally {
+                    Link head = link;
+                    while (head != null) {
+                        reclaimSpace(LINK_CAPACITY);
+                        head = head.next;
```

Comment:
@nitsanw should be free from GC nepotism here right?

Reply:
@johnou actually I think you have a good point here. We should ensure we null out next. Let me fix this

Reply:
@johnou PTAL again

Reply:
@normanmaurer :ok_hand: